libs/execution_context/ebpf_maps.c

```diff
@@ -444,33 +446,37 @@ _associate_inner_map(_Inout_ ebpf_core_object_map_t* object_map, ebpf_handle_t i
 
     // Convert value handle to an object pointer.
     result = ebpf_object_reference_by_handle(inner_map_handle, EBPF_OBJECT_MAP, &inner_map_template_object);
-    if (result != EBPF_SUCCESS)
+    if (result != EBPF_SUCCESS) {
+        EBPF_LOG_MESSAGE_NTSTATUS(
+            EBPF_TRACELOG_LEVEL_ERROR, EBPF_TRACELOG_KEYWORD_MAP, "Get object ref by handle failed.", result);
         goto Exit;
+    }
+
+    ebpf_core_map_t* template_core_map = (ebpf_core_map_t*)inner_map_template_object;
+    template_core_map = EBPF_FROM_FIELD(ebpf_core_map_t, object, inner_map_template_object);
 
-    object_map->inner_map_template = (ebpf_map_t*)inner_map_template_object;
-    object_map->core_map.ebpf_map_definition.inner_map_id = object_map->inner_map_template->object.id;
+    object_map->inner_template_map_definition = template_core_map->ebpf_map_definition;
+    object_map->core_map.ebpf_map_definition.inner_map_id = template_core_map->object.id;
 
 Exit:
+    if (inner_map_template_object) {
+        ebpf_object_release_reference(inner_map_template_object);
+    }
+
     return result;
 }
 
 static void
 _delete_object_array_map(_Inout_ _Post_invalid_ ebpf_core_map_t* map, ebpf_object_type_t value_type)
 {
-    ebpf_core_object_map_t* object_map = EBPF_FROM_FIELD(ebpf_core_object_map_t, core_map, map);
-
     // Release all entry references.
     for (uint32_t i = 0; i < map->ebpf_map_definition.max_entries; i++) {
         ebpf_id_t id = *(ebpf_id_t*)&map->data[i * map->ebpf_map_definition.value_size];
         if (id) {
-            ebpf_assert_success(ebpf_object_dereference_by_id(id, value_type));
+            ebpf_assert_success(ebpf_object_release_id_reference(id, value_type));
         }
     }
 
-    if (object_map->inner_map_template != NULL) {
-        ebpf_object_release_reference(&object_map->inner_map_template->object);
-    }
```

Comment:
Why is it we don't need to release the template id here?

Reply:
I'd noticed that we we were using the template map to only get its definition (to ensure each incoming map element's equivalence to the template map definition).  We now therefore store only the template map's definition (during the map-of-maps creation) and let go of the template map right then as I did not see any further dependence on the template map itself thereafter.

(I'd discussed this with @Alan-Jowett prior to going this route).